tools/flash_shelly.py

```diff
@@ -0,0 +1,499 @@
+    # logger.info('\nproperties: %s' % p_list)
+
+    device_list.sort()
+    for device in device_list:
+      probe_info(device + '.local', action, dry_run, silent_run, mode, exclude, exclude_device, forced_version, ffw)
+
+
+def usage():
+  print ("Shelly HomeKit flashing script utility")
+  print ("Usage: -{m|l|a|e|n|y|V|h} {hostname(s) optional}")
+  print (" -m {homekit|revert|keep}   Script mode.")
+  print (" -l            List info of shelly device.")
+  print (" -a            Run against all the devices on the network.")
+  print (" -e            Exclude hosts from found devices.")
+  print (" -n            Do a dummy run through.")
+  print (" -y            Do not ask any confirmation to perform the flash.")
+  print (" -V            Force a particular version.")
+  print (" -D            Enable Debug Logging, you can increase logging with (-DD).")
+  print (" -h            This help text.")
+
+def app(argv):
+  # Parse and interpret options.
+
+  try:
+    (opts, args) = getopt.getopt(argv[1:], ":aelnyhDm:V:")
+  except getopt.GetoptError as err:
+    logger.error(err)
+    usage()
+    exit(2)
+
+  action = "flash"
+  do_all = False
+  dry_run = False
+  silent_run = False
+  forced_version = False
+  exclude = False
+  mode = 'homekit'
+  ffw = None
+
+  for (opt, value) in opts:
+    if opt == "-m":
+      if value == 'homekit':
+        mode = 'homekit'
+      elif value == "revert":
+        mode='stock'
+      elif value == "keep":
+        mode="keep"
+      else:
+        print("Invalid option")
+        usage()
+        exit(2)
+    elif opt == "-a":
+      do_all = True
+    elif opt == "-e":
+      exclude = True
+    elif opt == "-l":
+      action = "list"
+    elif opt == "-n":
+      dry_run = True
+    elif opt == "-y":
+      silent_run = True
+    elif opt == "-V":
+      forced_version = True
+      ffw = value
+    elif opt == '-D':
+      if logger.getEffectiveLevel() >= 30:
+        logger.setLevel(logging.INFO)
+      else:
+        logger.setLevel(logging.DEBUG)
+    elif opt == '-h':
+      usage()
+      exit(0)
+
+  logger.info(WHITE + "app" + NC)
+  logger.info(PURPLE + "OS: %s\033[0m"% arch)
+  logger.info("ARG: %s" % argv)
+  logger.info("opts: %s" % opts)
+  logger.info("args: %s" % args)
+  logger.info("action: %s" % action)
+  logger.info("do_all: %s" % do_all)
+  logger.info("dry_run: %s" % dry_run)
+  logger.info("silent_run: %s" % silent_run)
+  logger.info("forced_version: %s" % forced_version)
+  logger.info("exclude: %s" % exclude)
+  logger.info("mode: %s" % mode)
+
+  if ((not opts and not args) and do_all == False) or (not args and do_all == False) or \
+     (args and do_all == True and exclude == False) or (not args and do_all == True and exclude == True):
+    usage()
+    exit(1)
+
+  try:
+    fp = urllib.request.urlopen("https://api.shelly.cloud/files/firmware")
+    global stock_release_info
+    stock_release_info = json.load(fp)
+    fp.close()
```

Comment:
instead of closing explicitly, use context manager:

with urllib.request.urlopen(...) as fp:
 ...

Reply:
done https://github.com/mongoose-os-apps/shelly-homekit/pull/176/commits/eeb8bae51752f3225a1425c7cec00adcb75675a3